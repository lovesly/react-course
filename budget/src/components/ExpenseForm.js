import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    // since we are using transform-class-properties, so we don't have to use 
    // constructor to access this.props.
    // what is the benifits to transform number to string??
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        note: this.props.expense ? this.props.expense.note : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
        calenderFocused: null,
        error: '',
    };
    onDescriptionChagne = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description })); 
    };
    onNoteChange = (e) => {
        // e.persist()
        const note = e.target.value;
        this.setState(() => ({ note })); 
    };
    onAmountChange = (e) => {
        // some input like 3.... or 12.12.... are not prevented at front end.
        // but some other operations are prevented. so fucking wierd.
        const amount = e.target.value;
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        // should trim the value.
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.'}));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,                
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            });
        }
    };
    render() {
        return (
            <form className="form" onSubmit={ this.onSubmit }>
                { this.state.error && <p className="form__error">{ this.state.error }</p> }
                <input
                    type="text"
                    className="text-input"
                    placeholder="Description"
                    autoFocus
                    value={ this.state.description }
                    onChange={ this.onDescriptionChagne }
                />
                <input 
                    type="number"
                    className="text-input"
                    placeholder="Amount"
                    value={ this.state.amount }
                    onChange={ this.onAmountChange }
                />
                <SingleDatePicker 
                    date={ this.state.createdAt }
                    onDateChange={ this.onDateChange }
                    focused={ this.state.calenderFocused }
                    onFocusChange={ this.onFocusChange }
                    numberOfMonths={ 1 }
                    isOutsideRange={ () => false }
                />
                <textarea 
                    className="textarea"
                    placeholder="Add a note for your expense (optional)"
                    value={ this.state.note }
                    onChange={ this.onNoteChange }
                >
                </textarea>
                <button className="button btn-4 btn-m">Save Expense</button>
            </form>
        );
    }
}