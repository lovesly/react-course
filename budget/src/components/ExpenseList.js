import React from 'react';
import { connect } from 'react-redux';

const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        <p>{ props.name }</p>
        <p>{ props.expenses[0].createdAt }</p>
        <p>{ props.filters.text}</p>
    </div>
);

const mapStateToProps = (state) => {
    console.log(state.filters);
    return {
        name: 'zz',
        expenses: state.expenses,
        filters: state.filters,
    };
};

export default connect(mapStateToProps)(ExpenseList);