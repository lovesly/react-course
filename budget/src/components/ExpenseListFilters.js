import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStart, setEnd } from '../actions/filters';

export class ExpenseListFilter extends React.Component {
    // why we use null here, and false in another place??
    // old question when will this.props be initialized?
    state = {
        calenderFocused: null,
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStart(startDate);
        this.props.setEnd(endDate);
    };
    onFocusChange = (focused) => {
        this.setState(() => ({ calenderFocused: focused }));
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        // console.log(e.target.value);
        switch (e.target.value) {
            case 'date':
                this.props.sortByDate();
                break;
            case 'amount':
                this.props.sortByAmount();
                break;
            default: 
                this.props.sortByDate();
        }
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            className="text-input"
                            placeholder="Search epxenses"
                            type="text" 
                            value={ this.props.filters.text } 
                            onChange={ this.onTextChange }
                        />
                    </div>
                    <div className="input-group__item">
                        <select 
                            className="text-input"                            
                            value={ this.props.filters.sortBy } 
                            onChange={ this.onSortChange }
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                            startDate={ this.props.filters.startDate }
                            startDateId="expense-list-filter-start"
                            endDate={ this.props.filters.endDate }
                            endDateId="expense-list-filter-end"
                            onDatesChange={ this.onDatesChange }
                            focusedInput={ this.state.calenderFocused }
                            onFocusChange={ this.onFocusChange }
                            showClearDates={ true }
                            numberOfMonths={ 1 }
                            isOutsideRange={ () => false }
                        />
                    </div>
                </div>
            </div>
        );
    };
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

const mapDispatchToProps = (dispatch) => ({ 
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStart: (startDate) => dispatch(setStart(startDate)),
    setEnd: (endDate) => dispatch(setEnd(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);