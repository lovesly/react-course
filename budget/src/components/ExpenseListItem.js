import React from 'react';
import { Link } from 'react-router-dom';

// where the fuck does dispatch come from?
const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <Link to={ `/edit/${ id }` }>
            <h3>{ description }</h3>
        </Link>
        <p>{ amount } - { createdAt }</p>
        <button onClick={() => {
            // what if I don't want to use link, I just like an edit button??? 
            // how can I access router here???
        }}>Edit</button>
    </div>
);

// if no mapStateToProps, how can we pass the params??
export default ExpenseListItem;




