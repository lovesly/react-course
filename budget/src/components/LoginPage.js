import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Budget</h1>
            <p>It's time to go, sly is waiting for me</p>
            <button className="button btn-4" onClick={ startLogin }><span>Login with Google</span></button>
        </div>
    </div>

);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
