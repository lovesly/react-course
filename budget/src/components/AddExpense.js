import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses'

// to test the unconnected version, export the class.
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => { 
        // ? why this one is easier to test?
        // props.dispatch(addExpense(expense));
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>    
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        onSubmit={ this.onSubmit }
                    />
                </div>
            </div>
        );
    }
}

// to avoid inline functions? so use class based component??? wtf dude.
const mapDispatchToProps = (dispatch) => ({ 
    startAddExpense: (expense) => dispatch(startAddExpense(expense)) 
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);