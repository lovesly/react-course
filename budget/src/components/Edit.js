import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

// change from stateless to class,so we don't have to redefine the inline functions 
// every single time the component gets rendered. 
export class EditPage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onClick = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>    
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={ this.props.expense }
                        onSubmit={ this.onSubmit }
                    />
                    <button className="button btn-4 btn-logout btn-m" onClick={ this.onClick }>Remove</button>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

// in real env, props is not defined??
const mapDispatchToProps = (dispatch) => ({ 
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense({ id })),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPage);