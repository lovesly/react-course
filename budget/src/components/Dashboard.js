import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilters';

const BudgetDashboardPage = () => (
    <div>
        <ExpenseListFilter />
        <ExpenseList />
    </div>
);

export default BudgetDashboardPage;