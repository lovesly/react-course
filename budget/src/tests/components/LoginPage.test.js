import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

let startLogin, wrapper;
beforeEach(() => {
    startLogin = jest.fn();
    // history = { push: jest.fn() };
    wrapper = shallow(<LoginPage startLogin={ startLogin } />);
});

test(`should render login page correctly`, () => {
    expect(wrapper).toMatchSnapshot();
});

test(`should call startLogin on button click`, () => {
    wrapper.find('button').simulate('click');
    // expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startLogin).toHaveBeenCalled();      
});