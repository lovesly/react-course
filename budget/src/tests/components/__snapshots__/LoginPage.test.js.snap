// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render login page correctly 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Budget
    </h1>
    <p>
      It's time to go, sly is waiting for me
    </p>
    <button
      className="button btn-4"
      onClick={[MockFunction]}
    >
      <span>
        Login with Google
      </span>
    </button>
  </div>
</div>
`;
